src/dids.ts

```diff
@@ -48,32 +48,47 @@ export async function generateDid({ signingPublicJwk, encryptionPublicJwk, recov
         return [revealValueEncodedString, commitmentHashHashEncodedString];
     };
 
+    const publicKeyEntry = (id: string, jwk: any, additionalPurposes: string[] = []) => {
+        return {
+            id,
+            purposes: ["assertionMethod", ...additionalPurposes],
+            type: "JsonWebKey2020",
+            publicKeyJwk: JSON.parse(JSON.stringify({ ...jwk, kid: undefined }))
+        }
+    };
+
     const [recoveryValue, recoveryCommitment] = revealCommitPair(recoveryPublicJwk);
     const [updateValue, updateCommitment] = revealCommitPair(updatePublicJwk);
-    let patches: {action: string, publicKeys?: any, services?: any}[] = [{
-            action: 'add-public-keys',
-            publicKeys: [{
-                id: 'signing-key-1',
-                purposes: ['authentication', 'assertionMethod'],
-                type: 'JsonWebKey2020',
-                publicKeyJwk: JSON.parse(JSON.stringify({...signingPublicJwk, kid: undefined}))
-            }, {
-                id: 'encryption-key-1',
-                purposes: ['keyAgreement', 'assertionMethod'],
-                type: 'JsonWebKey2020',
-                publicKeyJwk: JSON.parse(JSON.stringify({...encryptionPublicJwk, kid: undefined}))
-            }]
-        }]
+    const publicKeys = [];
+
+    if (typeof signingPublicJwk === "object") {
+        publicKeys.push(publicKeyEntry("signing-key-1", signingPublicJwk, ["authentication"]));
+    }
+    if (typeof encryptionPublicJwk === "object") {
+        publicKeys.push(publicKeyEntry("encryption-key-1", encryptionPublicJwk, ["keyAgreement"]));
+    }
+
+    const patches: { action: string, publicKeys?: any, services?: any }[] = [];
+
+    if (publicKeys.length > 0) {
+        patches.push({
+            action: "add-public-keys",
+            publicKeys
+        });
+    }
+
     if (domains.length > 0) {
         patches.push({
             "action": "add-services",
-            "services": [{
-                "id": "linked-domain",
-                "type": "LinkedDomains",
-                "serviceEndpoint": domains[0]
-                }]
-            });
-    } 
+            "services": domains.map((domain, index) => {
+                return {
+                    "id": `linked-domain-${index + 1}`,
+                    "type": "LinkedDomains",
+                    "serviceEndpoint": domain
```

Comment:
Technically there's an option to provide an array of origins within a single `LinkedDomains` entry in the `services[]` array, but I think this "exploded" list is cleaner / more likely to be compatible with client implementations (and ... we only have a use case for one right now anyway)